arangod/Aql/AqlItemMatrix.cpp

```diff
@@ -48,8 +50,41 @@ std::vector<AqlItemMatrix::RowIndex> AqlItemMatrix::produceRowIndexes() const {
     result.reserve(size());
     uint32_t index = 0;
     for (auto const& block : _blocks) {
-      for (uint32_t row = 0; row < block->size(); ++row) {
-        result.emplace_back(index, row);
+      // Default case, 0 -> end
+      uint32_t startRow = 0;
+      uint32_t endRow = block->size();
+
+      if (block == _blocks.front() && block->hasShadowRows()) {
```

Comment:
I wonder whether this could be done sligthly more readable if one handled the first and last blocks outside of the loop? This might be a personal preference thing though, and I did not try rewriting this code just now.

Reply:
i am a bit unhappy with this loop as well yes.
Let me check what I can do